include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -104,6 +130,50 @@ class cuda_vm_resource : public memory_resource<memory_kind::device> {
     stat_ = {};
   }
 
+ protected:
+  bool defer_dealloc_ = false;
+  constexpr bool deferred_dealloc_enabled() const noexcept { return defer_dealloc_; }
+  static constexpr int deferred_dealloc_max_outstanding() { return 16; }
+
+  virtual void flush_deferred() {}
+
+  void do_deallocate(void *ptr, size_t size, size_t alignment) override {
+    deallocate_impl(ptr, size, alignment, true, true);
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    if (size == 0)
+      return nullptr;
+    adjust_params(size, alignment);
+    std::unique_lock<pool_lock_t> pool_guard(pool_lock_);
+    // try to get a free region that's already mapped
+    void *ptr = try_get_mapped(size, alignment);
+    if (ptr) {
+      stat_add_allocation(size);
+      return ptr;
+    }
+    if (deferred_dealloc_enabled()) {
+      pool_guard.unlock();
+      flush_deferred();
+      pool_guard.lock();
+      ptr = try_get_mapped(size, alignment);
+      if (ptr) {
+        stat_add_allocation(size);
+        return ptr;
+      }
+    }
+    DeviceGuard dg(device_ordinal_);
+    mem_lock_guard mem_guard(mem_lock_);
+    void *va = get_va(size, alignment);
+    map_storage(va, size);
+    ptr = free_mapped_.get_specific_block(va, size);
+    assert(ptr == va);
+    stat_take_free(size);
+    mark_as_unavailable(ptr, size);
```

Comment:
This was the bug - previously this code was ```ptr = try_get_mapped(size, alignment)``` - if somehow it got a block that wasn't entirely covered by `va` there was inconsistency in what's free and what is not and the integrity of the pool was destroyed. The bug was exposed when `flush_deferred()` wasn't followed by another attempt to get mapped already memory.